crypto/cmp/cmp_ctx.c

```diff
@@ -0,0 +1,1276 @@
+    ctx->clCert = NULL;
+
+    if ((ctx->clCert = X509_dup(cert)) == NULL) {
+        CMPerr(CMP_F_OSSL_CMP_CTX_SET1_CLCERT, ERR_R_MALLOC_FAILURE);
+        return 0;
+    }
+    return 1;
+ err:
+    CMPerr(CMP_F_OSSL_CMP_CTX_SET1_CLCERT, CMP_R_NULL_ARGUMENT);
+    return 0;
+}
+
+/*
+ * Set the old certificate that we are updating in KUR
+ * or the certificate to be revoked in RR, respectively.
+ * Also used as reference cert (defaulting to clCert) for deriving subject DN
+ * and SANs. Its issuer is used as default recipient in the CMP message header.
+ * returns 1 on success, 0 on error
+ */
+int OSSL_CMP_CTX_set1_oldClCert(OSSL_CMP_CTX *ctx, const X509 *cert)
+{
+    if (ctx == NULL || cert == NULL)
+        goto err;
+
+    X509_free(ctx->oldClCert);
+    ctx->oldClCert = NULL;
+
+    if ((ctx->oldClCert = X509_dup(cert)) == NULL) {
+        CMPerr(CMP_F_OSSL_CMP_CTX_SET1_OLDCLCERT, ERR_R_MALLOC_FAILURE);
+        return 0;
+    }
+    return 1;
+ err:
+    CMPerr(CMP_F_OSSL_CMP_CTX_SET1_OLDCLCERT, CMP_R_NULL_ARGUMENT);
+    return 0;
+}
+
+/*
+ * Set the PKCS#10 CSR to be sent in P10CR
+ * returns 1 on success, 0 on error
+ */
+int OSSL_CMP_CTX_set1_p10CSR(OSSL_CMP_CTX *ctx, const X509_REQ *csr)
+{
+    if (ctx == NULL || csr == NULL)
+        goto err;
+
+    X509_REQ_free(ctx->p10CSR);
+    ctx->p10CSR = NULL;
+
+    if ((ctx->p10CSR = X509_REQ_dup(csr)) == NULL) {
+        CMPerr(CMP_F_OSSL_CMP_CTX_SET1_P10CSR, ERR_R_MALLOC_FAILURE);
+        return 0;
+    }
+    return 1;
+ err:
+    CMPerr(CMP_F_OSSL_CMP_CTX_SET1_P10CSR, CMP_R_NULL_ARGUMENT);
+    return 0;
+}
+
+/*
+ * sets the (newly received in IP/KUP/CP) client Certificate to the context
+ * returns 1 on success, 0 on error
+ * TODO: this only permits for one client cert to be received...
+ */
+int CMP_CTX_set1_newClCert(OSSL_CMP_CTX *ctx, X509 *cert)
+{
+    if (ctx == NULL || cert == NULL)
+        goto err;
+
+    X509_free(ctx->newClCert);
+    ctx->newClCert = NULL;
+
+    if (!X509_up_ref(cert))
+        goto err;
+    ctx->newClCert = cert;
+    return 1;
+ err:
+    CMPerr(CMP_F_CMP_CTX_SET1_NEWCLCERT, CMP_R_NULL_ARGUMENT);
+    return 0;
+}
+
+/*
+ * Get the (newly received in IP/KUP/CP) client certificate from the context
+ * TODO: this only permits for one client cert to be received...
+ */
+X509 *CMP_CTX_get0_newClCert(const OSSL_CMP_CTX *ctx)
+{
+    if (ctx == NULL)
+        return NULL;
+    return ctx->newClCert;
+}
+
+/*
+ * Set the client's private key. This creates a duplicate of the key
+ * so the given pointer is not used directly.
+ * returns 1 on success, 0 on error
+ */
+int OSSL_CMP_CTX_set1_pkey(OSSL_CMP_CTX *ctx, EVP_PKEY *pkey)
+{
+    if (ctx == NULL || pkey == NULL)
+        goto err;
+
+    if (!EVP_PKEY_up_ref(pkey))
+        return 0;
+    if (OSSL_CMP_CTX_set0_pkey(ctx, pkey))
+        return 1;
+    EVP_PKEY_free(pkey); /* down ref */
+    return 0;
+
+ err:
+    CMPerr(CMP_F_OSSL_CMP_CTX_SET1_PKEY, CMP_R_NULL_ARGUMENT);
```

Comment:
I like the idea of feeding the error queue consistently in all error paths.
I've implemented this now in `cmp_ctx.c` as far as possible.

Doing so I followed various invocation paths into lower-level functions of the crypto lib and found that many of the functions there are inconsistent in this regard, including most of the functions (indirectly) used here, such as `ASN1_STRING_set`,  `OPENSSL_sk_dup`, and `OBJ_nid2obj`. So the higher-level CMP functions would have a pretty hard time trying to be fully consistent w.r.t the error queue for this reason.

While the suggestion to add comments like "CMPerr already called" sounds reasonable at first, this would be pretty tedious to do and add quite some clutter. I guess this is why it is not done elsewhere in the crypto lib, either. 

So my suggestion is to adopt the following convention: each function is responsible for consistency w.r.t. the error queue only within its own body, and should (be allowed to) rely on the same consistency for all lower-level function it calls.